File: ports/python/src/worldcut/verifier.py
from __future__ import annotations

from typing import Literal, cast

from .canonicalization import sha256_digest, utf16_sort_key
from .errors import WorldCutInputError
from .models import (
    CommonValidTimeRequirement,
    ContractVerdict,
    DependencyRequirement,
    JsonValue,
    ParsedInput,
    RequirementResult,
    ValueEqualsRequirement,
    VerificationCoverage,
    VerificationResult,
    _unwrap_parsed_input,
    thaw_json,
)
from .planning import select_acquisition_plan
from .requirements import (
    evaluate_common_valid_time,
    evaluate_dependency,
    evaluate_value_equals,
)
from .validation import parse_input

ENGINE_VERSION = "0.1.2"
CANONICALIZATION: Literal["worldcut-json-v1"] = "worldcut-json-v1"


def verify(parsed_input: ParsedInput) -> VerificationResult:
    """Verify a parsed WorldCut input and return a fresh mutable result."""

    try:
        value = _unwrap_parsed_input(parsed_input)
    except TypeError as error:
        raise WorldCutInputError(str(error)) from error
    observations = {observation.role: observation for observation in value.observations}
    requirements = sorted(
        value.contract.requirements, key=lambda item: utf16_sort_key(item.id)
    )
    results: list[RequirementResult] = []
    for requirement in requirements:
        if isinstance(requirement, DependencyRequirement):
            results.append(evaluate_dependency(requirement, observations))
        elif isinstance(requirement, CommonValidTimeRequirement):
            results.append(evaluate_common_valid_time(requirement, observations))
        elif isinstance(requirement, ValueEqualsRequirement):
            results.append(evaluate_value_equals(requirement, observations))

    coverage: VerificationCoverage = {
        "required": 0,
        "satisfied": 0,
        "violated": 0,
        "unknown": 0,
        "advisory": 0,
    }
    for result in results:
        if not result["required"]:
            coverage["advisory"] += 1
            continue
        coverage["required"] += 1
        if result["status"] == "SATISFIED":
            coverage["satisfied"] += 1
        elif result["status"] == "VIOLATED":
            coverage["violated"] += 1
        else:
            coverage["unknown"] += 1

    verdict: ContractVerdict
    if coverage["violated"]:
        verdict = "CONTRACT_VIOLATED"
    elif coverage["unknown"]:
        verdict = "INSUFFICIENT_EVIDENCE"
    else:
        verdict = "CONTRACT_SATISFIED"
    plan = select_acquisition_plan(results)

    contract = cast(dict[str, JsonValue], thaw_json(value.contract.raw))
    contract["requirements"] = [
        thaw_json(requirement.raw) for requirement in requirements
    ]
    record: JsonValue = {
        "protocolVersion": value.protocol_version,
        "engineVersion": ENGINE_VERSION,
        "canonicalization": CANONICALIZATION,
        "contract": contract,
        "observations": [
            thaw_json(observation.raw)
            for observation in sorted(
                value.observations, key=lambda item: utf16_sort_key(item.role)
            )
        ],
        "verdict": verdict,
        "requirementResults": cast(list[JsonValue], results),
        "acquisitionPlan": cast(JsonValue, plan),
    }
    return {
        "protocolVersion": value.protocol_version,
        "engineVersion": ENGINE_VERSION,
        "canonicalization": CANONICALIZATION,
        "contractId": value.contract.id,
        "contractVersion": value.contract.version,
        "verdict": verdict,
        "coverage": coverage,
        "requirementResults": results,
        "acquisitionPlan": plan,
        "verificationRecordDigest": sha256_digest(record),
    }


def verify_json(source: str | bytes | bytearray) -> VerificationResult:
    """Parse and verify one WorldCut JSON document."""

    return verify(parse_input(source))

File: ports/go/verifier.go
package worldcut

import (
	"encoding/json"
	"fmt"
	"sort"
)

func VerifyJSON(source []byte) (*VerificationResult, error) {
	input, err := ParseInput(source)
	if err != nil {
		return nil, err
	}
	return Verify(input)
}

func Verify(input *ParsedInput) (*VerificationResult, error) {
	if input == nil || input.value == nil || input.value.raw == nil {
		return nil, invalidInput("verification input must be produced by ParseInput")
	}
	return verifyParsed(input.value)
}

func verifyParsed(input *verificationInput) (*VerificationResult, error) {
	observations := make(map[string]*Observation, len(input.Observations))
	for i := range input.Observations {
		observation := &input.Observations[i]
		observations[observation.Role] = observation
	}
	requirements := append([]Requirement{}, input.Contract.Requirements...)
	sort.Slice(requirements, func(i, j int) bool {
		return compareUTF16(requirements[i].ID, requirements[j].ID) < 0
	})
	results := make([]RequirementResult, 0, len(requirements))
	for _, requirement := range requirements {
		var result RequirementResult
		var err error
		switch requirement.Type {
		case "dependency":
			result, err = evaluateDependency(requirement, observations)
		case "common_valid_time":
			result, err = evaluateCommonValidTime(requirement, observations)
		case "value_equals":
			result, err = evaluateValueEquals(requirement, observations)
		default:
			err = fmt.Errorf("unsupported requirement type: %s", requirement.Type)
		}
		if err != nil {
			return nil, err
		}
		results = append(results, result)
	}
	coverage := VerificationCoverage{}
	for _, result := range results {
		if !result.Required {
			coverage.Advisory++
			continue
		}
		coverage.Required++
		switch result.Status {
		case "SATISFIED":
			coverage.Satisfied++
		case "VIOLATED":
			coverage.Violated++
		case "UNKNOWN":
			coverage.Unknown++
		}
	}
	verdict := "CONTRACT_SATISFIED"
	if coverage.Violated > 0 {
		verdict = "CONTRACT_VIOLATED"
	} else if coverage.Unknown > 0 {
		verdict = "INSUFFICIENT_EVIDENCE"
	}
	plan, err := SelectAcquisitionPlan(results)
	if err != nil {
		return nil, err
	}
	record := map[string]any{
		"protocolVersion":    input.ProtocolVersion,
		"engineVersion":      EngineVersion,
		"canonicalization":   CanonicalizationName,
		"contract":           normalizedContract(input),
		"observations":       normalizedObservations(input),
		"verdict":            verdict,
		"requirementResults": results,
		"acquisitionPlan":    plan,
	}
	digest, err := SHA256Digest(record)
	if err != nil {
		return nil, err
	}
	result := &VerificationResult{
		ProtocolVersion:          input.ProtocolVersion,
		EngineVersion:            EngineVersion,
		Canonicalization:         CanonicalizationName,
		ContractID:               input.Contract.ID,
		ContractVersion:          input.Contract.Version,
		Verdict:                  verdict,
		Coverage:                 coverage,
		RequirementResults:       results,
		AcquisitionPlan:          plan,
		VerificationRecordDigest: digest,
	}
	return cloneVerificationResult(result)
}

func cloneVerificationResult(result *VerificationResult) (*VerificationResult, error) {
	encoded, err := json.Marshal(result)
	if err != nil {
		return nil, err
	}
	var cloned VerificationResult
	if err := json.Unmarshal(encoded, &cloned); err != nil {
		return nil, err
	}
	return &cloned, nil
}

func normalizedContract(input *verificationInput) map[string]any {
	contract := make(map[string]any, len(input.Contract.raw))
	for key, value := range input.Contract.raw {
		contract[key] = value
	}
	requirements := append([]Requirement{}, input.Contract.Requirements...)
	sort.Slice(requirements, func(i, j int) bool {
		return compareUTF16(requirements[i].ID, requirements[j].ID) < 0
	})
	rawRequirements := make([]any, len(requirements))
	for i, requirement := range requirements {
		rawRequirements[i] = requirement.raw
	}
	contract["requirements"] = rawRequirements
	return contract
}

func normalizedObservations(input *verificationInput) []any {
	observations := append([]Observation{}, input.Observations...)
	sort.Slice(observations, func(i, j int) bool {
		return compareUTF16(observations[i].Role, observations[j].Role) < 0
	})
	result := make([]any, len(observations))
	for i, observation := range observations {
		result[i] = observation.raw
	}
	return result
}
